File: nasdaq-cse/market_data/provider.py
"""
Market data module for fetching real-time gold prices and generating charts
"""
import asyncio
import json
import random
import time
from datetime import datetime, timedelta
from typing import List, Dict, Optional
import requests
import pandas as pd
import plotly.graph_objects as go
import plotly.utils
from plotly.subplots import make_subplots


class GoldPriceProvider:
    """
    Provides real-time gold price data from multiple sources
    """
    
    def __init__(self):
        self.base_price = 2050.0  # Base gold price in USD per ounce
        self.current_price = self.base_price
        self.price_history = []
        self.last_update = datetime.utcnow()
    
    async def fetch_real_gold_price(self) -> Optional[float]:
        """
        Fetch real gold price from external API
        Falls back to simulation if API is unavailable
        """
        try:
            # Try to fetch from a free gold price API
            response = requests.get(
                "https://api.metals.live/v1/spot/gold",
                timeout=5
            )
            if response.status_code == 200:
                data = response.json()
                return float(data.get('price', self.current_price))
        except Exception as e:
            print(f"Failed to fetch real gold price: {e}")
        
        # Fallback to simulation
        return self.simulate_price_movement()
    
    def simulate_price_movement(self) -> float:
        """
        Simulate realistic gold price movements
        """
        # Calculate time since last update
        now = datetime.utcnow()
        time_diff = (now - self.last_update).total_seconds()
        
        # Random walk with mean reversion
        volatility = 0.02  # 2% volatility
        mean_reversion = 0.001
        drift = -mean_reversion * (self.current_price - self.base_price) / self.base_price
        
        # Random component
        random_component = random.gauss(0, volatility) * (time_diff / 3600) ** 0.5
        
        # Calculate price change
        price_change = drift + random_component
        self.current_price *= (1 + price_change)
        
        # Add some noise for intraday movements
        self.current_price += random.gauss(0, 0.5)
        
        # Keep price within reasonable bounds
        self.current_price = max(1800, min(2500, self.current_price))
        
        self.last_update = now
        return self.current_price
    
    async def get_current_price(self) -> Dict:
        """
        Get current gold price with bid/ask spread
        """
        price = await self.fetch_real_gold_price()
        spread = random.uniform(0.5, 1.5)  # Bid-ask spread
        
        price_data = {
            'timestamp': datetime.utcnow().isoformat(),
            'price': round(price, 2),
            'bid': round(price - spread/2, 2),
            'ask': round(price + spread/2, 2),
            'volume': random.randint(100, 1000),
            'change_24h': round(random.uniform(-2.0, 2.0), 2),
            'change_percent': round(random.uniform(-0.1, 0.1), 4)
        }
        
        # Store in history
        self.price_history.append(price_data)
        
        # Keep only last 1000 price points
        if len(self.price_history) > 1000:
            self.price_history = self.price_history[-1000:]
        
        return price_data
    
    def get_price_history(self, hours: int = 24) -> List[Dict]:
        """
        Get historical price data for specified hours
        """
        cutoff_time = datetime.utcnow() - timedelta(hours=hours)
        return [
            price for price in self.price_history
            if datetime.fromisoformat(price['timestamp']) > cutoff_time
        ]


class ChartGenerator:
    """
    Generates interactive charts for market data visualization
    """
    
    def __init__(self, price_provider: GoldPriceProvider):
        self.price_provider = price_provider
    
    def create_price_chart(self, hours: int = 24) -> str:
        """
        Create an interactive price chart using Plotly
        """
        price_history = self.price_provider.get_price_history(hours)
        
        if not price_history:
            return self._create_empty_chart()
        
        # Convert to DataFrame for easier manipulation
        df = pd.DataFrame(price_history)
        df['timestamp'] = pd.to_datetime(df['timestamp'])
        
        # Create candlestick-style chart
        fig = make_subplots(
            rows=3, cols=1,
            shared_xaxes=True,
            vertical_spacing=0.02,
            subplot_titles=('Gold Price (USD/oz)', 'Volume', 'Price Change %'),
            row_heights=[0.6, 0.2, 0.2]
        )
        
        # Price line chart
        fig.add_trace(
            go.Scatter(
                x=df['timestamp'],
                y=df['price'],
                mode='lines',
                name='Price',
                line=dict(color='gold', width=2),
                hovertemplate='<b>Price</b>: $%{y:.2f}<br><b>Time</b>: %{x}<extra></extra>'
            ),
            row=1, col=1
        )
        
        # Bid-Ask spread
        fig.add_trace(
            go.Scatter(
                x=df['timestamp'],
                y=df['ask'],
                mode='lines',
                name='Ask',
                line=dict(color='red', width=1, dash='dash'),
                showlegend=False
            ),
            row=1, col=1
        )
        
        fig.add_trace(
            go.Scatter(
                x=df['timestamp'],
                y=df['bid'],
                mode='lines',
                name='Bid',
                line=dict(color='green', width=1, dash='dash'),
                fill='tonexty',
                fillcolor='rgba(255,215,0,0.1)',
                showlegend=False
            ),
            row=1, col=1
        )
        
        # Volume chart
        fig.add_trace(
            go.Bar(
                x=df['timestamp'],
                y=df['volume'],
                name='Volume',
                marker_color='lightblue',
                showlegend=False
            ),
            row=2, col=1
        )
        
        # Price change percentage
        fig.add_trace(
            go.Scatter(
                x=df['timestamp'],
                y=df['change_percent'],
                mode='lines+markers',
                name='Change %',
                line=dict(color='purple', width=1),
                marker=dict(size=3),
                showlegend=False
            ),
            row=3, col=1
        )
        
        # Update layout
        fig.update_layout(
            title='Gold Derivatives Market Data - Live Feed',
            xaxis_title='Time',
            height=600,
            template='plotly_white',
            showlegend=True,
            legend=dict(
                orientation="h",
                yanchor="bottom",
                y=1.02,
                xanchor="right",
                x=1
            )
        )
        
        fig.update_yaxes(title_text="Price (USD)", row=1, col=1)
        fig.update_yaxes(title_text="Volume", row=2, col=1)
        fig.update_yaxes(title_text="Change %", row=3, col=1)
        
        return fig.to_json()
    
    def create_pnl_chart(self, positions_data: List[Dict]) -> str:
        """
        Create P&L chart for position tracking
        """
        if not positions_data:
            return self._create_empty_chart()
        
        df = pd.DataFrame(positions_data)
        
        fig = go.Figure()
        
        # Unrealized P&L
        fig.add_trace(
            go.Scatter(
                x=df['timestamp'],
                y=df['unrealized_pnl'],
                mode='lines+markers',
                name='Unrealized P&L',
                line=dict(color='blue', width=2),
                marker=dict(size=5)
            )
        )
        
        # Realized P&L
        fig.add_trace(
            go.Scatter(
                x=df['timestamp'],
                y=df['realized_pnl'],
                mode='lines+markers',
                name='Realized P&L',
                line=dict(color='green', width=2),
                marker=dict(size=5)
            )
        )
        
        # Add zero line
        fig.add_hline(y=0, line_dash="dash", line_color="gray")
        
        fig.update_layout(
            title='Position P&L Tracking',
            xaxis_title='Time',
            yaxis_title='P&L (USD)',
            height=400,
            template='plotly_white'
        )
        
        return fig.to_json()
    
    def create_exposure_chart(self, exposure_data: Dict) -> str:
        """
        Create exposure analysis chart
        """
        contracts = list(exposure_data.keys())
        exposures = list(exposure_data.values())
        
        fig = go.Figure()
        
        fig.add_trace(
            go.Bar(
                x=contracts,
                y=exposures,
                marker_color=['red' if x < 0 else 'green' for x in exposures],
                text=[f"${x:,.0f}" for x in exposures],
                textposition='auto'
            )
        )
        
        fig.update_layout(
            title='Position Exposure by Contract',
            xaxis_title='Contract',
            yaxis_title='Exposure (USD)',
            height=300,
            template='plotly_white'
        )
        
        return fig.to_json()
    
    def _create_empty_chart(self) -> str:
        """
        Create an empty chart when no data is available
        """
        fig = go.Figure()
        fig.add_annotation(
            text="No data available",
            xref="paper", yref="paper",
            x=0.5, y=0.5, xanchor='center', yanchor='middle',
            showarrow=False,
            font=dict(size=20, color="gray")
        )
        fig.update_layout(
            height=400,
            template='plotly_white',
            xaxis=dict(visible=False),
            yaxis=dict(visible=False)
        )
        return fig.to_json()


# Global instances
gold_price_provider = GoldPriceProvider()
chart_generator = ChartGenerator(gold_price_provider)

File: nasdaq-cse-go/internal/marketdata/provider.go
// Package marketdata provides market data functionality for gold prices and charts
package marketdata

import (
	"encoding/json"
	"fmt"
	"math"
	"math/rand"
	"net/http"
	"sync"
	"time"

	"github.com/smaruf/python-ai-course/nasdaq-cse-go/internal/core"
)

// GoldPriceProvider provides real-time gold price data
type GoldPriceProvider struct {
	basePrice    float64
	currentPrice float64
	priceHistory []core.MarketDataResponse
	lastUpdate   time.Time
	mutex        sync.RWMutex
}

// NewGoldPriceProvider creates a new gold price provider
func NewGoldPriceProvider() *GoldPriceProvider {
	return &GoldPriceProvider{
		basePrice:    2050.0,
		currentPrice: 2050.0,
		priceHistory: make([]core.MarketDataResponse, 0),
		lastUpdate:   time.Now(),
	}
}

// ExternalGoldPrice represents response from external gold price API
type ExternalGoldPrice struct {
	Price float64 `json:"price"`
}

// FetchRealGoldPrice attempts to fetch real gold price from external API
func (gpp *GoldPriceProvider) FetchRealGoldPrice() (float64, error) {
	// Try to fetch from a free gold price API
	client := &http.Client{Timeout: 5 * time.Second}
	resp, err := client.Get("https://api.metals.live/v1/spot/gold")
	if err != nil {
		return 0, fmt.Errorf("failed to fetch gold price: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return 0, fmt.Errorf("API returned status code: %d", resp.StatusCode)
	}

	var goldPrice ExternalGoldPrice
	if err := json.NewDecoder(resp.Body).Decode(&goldPrice); err != nil {
		return 0, fmt.Errorf("failed to decode API response: %w", err)
	}

	return goldPrice.Price, nil
}

// SimulatePriceMovement simulates realistic gold price movements
func (gpp *GoldPriceProvider) SimulatePriceMovement() float64 {
	gpp.mutex.Lock()
	defer gpp.mutex.Unlock()

	now := time.Now()
	timeDiff := now.Sub(gpp.lastUpdate).Seconds()

	// Random walk with mean reversion
	volatility := 0.02 // 2% volatility
	meanReversion := 0.001
	drift := -meanReversion * (gpp.currentPrice - gpp.basePrice) / gpp.basePrice

	// Random component
	randomComponent := rand.NormFloat64() * volatility * math.Sqrt(timeDiff/3600)

	// Calculate price change
	priceChange := drift + randomComponent
	gpp.currentPrice *= (1 + priceChange)

	// Add some noise for intraday movements
	gpp.currentPrice += rand.NormFloat64() * 0.5

	// Keep price within reasonable bounds
	if gpp.currentPrice < 1800 {
		gpp.currentPrice = 1800
	}
	if gpp.currentPrice > 2500 {
		gpp.currentPrice = 2500
	}

	gpp.lastUpdate = now
	return gpp.currentPrice
}

// GetCurrentPrice returns current gold price with bid/ask spread
func (gpp *GoldPriceProvider) GetCurrentPrice() core.MarketDataResponse {
	// Try to fetch real price, fallback to simulation
	price, err := gpp.FetchRealGoldPrice()
	if err != nil {
		price = gpp.SimulatePriceMovement()
	} else {
		gpp.mutex.Lock()
		gpp.currentPrice = price
		gpp.mutex.Unlock()
	}

	// Calculate bid-ask spread
	spread := rand.Float64()*1.0 + 0.5 // 0.5 to 1.5 spread

	priceData := core.MarketDataResponse{
		Timestamp:     time.Now(),
		Price:         math.Round(price*100) / 100,
		Bid:           math.Round((price-spread/2)*100) / 100,
		Ask:           math.Round((price+spread/2)*100) / 100,
		Volume:        int64(rand.Intn(900) + 100), // 100-1000 volume
		Change24h:     math.Round((rand.Float64()*4.0-2.0)*100) / 100, // -2.0 to +2.0
		ChangePercent: math.Round((rand.Float64()*0.2-0.1)*10000) / 10000, // -0.1 to +0.1
	}

	// Store in history
	gpp.mutex.Lock()
	gpp.priceHistory = append(gpp.priceHistory, priceData)
	
	// Keep only last 1000 price points
	if len(gpp.priceHistory) > 1000 {
		gpp.priceHistory = gpp.priceHistory[len(gpp.priceHistory)-1000:]
	}
	gpp.mutex.Unlock()

	return priceData
}

// GetPriceHistory returns historical price data for specified duration
func (gpp *GoldPriceProvider) GetPriceHistory(hours int) []core.MarketDataResponse {
	gpp.mutex.RLock()
	defer gpp.mutex.RUnlock()

	cutoffTime := time.Now().Add(-time.Duration(hours) * time.Hour)
	var result []core.MarketDataResponse

	for _, price := range gpp.priceHistory {
		if price.Timestamp.After(cutoffTime) {
			result = append(result, price)
		}
	}

	return result
}

// ChartGenerator generates chart data for market visualization
type ChartGenerator struct {
	priceProvider *GoldPriceProvider
}

// NewChartGenerator creates a new chart generator
func NewChartGenerator(provider *GoldPriceProvider) *ChartGenerator {
	return &ChartGenerator{priceProvider: provider}
}

// CreatePriceChartData creates price chart data
func (cg *ChartGenerator) CreatePriceChartData(hours int) core.ChartDataResponse {
	priceHistory := cg.priceProvider.GetPriceHistory(hours)
	
	if len(priceHistory) == 0 {
		return core.ChartDataResponse{
			Data: []core.ChartDataPoint{},
			Type: "price",
		}
	}

	// Convert to chart data points
	var chartData []core.ChartDataPoint
	for _, price := range priceHistory {
		chartData = append(chartData, core.ChartDataPoint{
			Timestamp: price.Timestamp,
			Price:     price.Price,
			Volume:    price.Volume,
		})
	}

	return core.ChartDataResponse{
		Data: chartData,
		Type: "price",
	}
}

// CreatePnLChartData creates P&L chart data for positions
func (cg *ChartGenerator) CreatePnLChartData(positionsData []map[string]interface{}) core.ChartDataResponse {
	if len(positionsData) == 0 {
		return core.ChartDataResponse{
			Data: []core.ChartDataPoint{},
			Type: "pnl",
		}
	}

	var chartData []core.ChartDataPoint
	for _, pos := range positionsData {
		timestamp, _ := time.Parse(time.RFC3339, pos["timestamp"].(string))
		unrealizedPnL, _ := pos["unrealized_pnl"].(float64)
		
		chartData = append(chartData, core.ChartDataPoint{
			Timestamp: timestamp,
			Price:     unrealizedPnL,
			Volume:    0, // Not applicable for P&L
		})
	}

	return core.ChartDataResponse{
		Data: chartData,
		Type: "pnl",
	}
}

// CreateExposureChartData creates exposure analysis chart data
func (cg *ChartGenerator) CreateExposureChartData(exposureData map[string]float64) map[string]interface{} {
	contracts := make([]string, 0)
	exposures := make([]float64, 0)

	for contract, exposure := range exposureData {
		contracts = append(contracts, contract)
		exposures = append(exposures, exposure)
	}

	return map[string]interface{}{
		"contracts": contracts,
		"exposures": exposures,
		"type":      "exposure",
	}
}

// MarketDataService provides market data functionality
type MarketDataService struct {
	priceProvider  *GoldPriceProvider
	chartGenerator *ChartGenerator
}

// NewMarketDataService creates a new market data service
func NewMarketDataService() *MarketDataService {
	provider := NewGoldPriceProvider()
	generator := NewChartGenerator(provider)

	return &MarketDataService{
		priceProvider:  provider,
		chartGenerator: generator,
	}
}

// GetCurrentPrice returns current market price
func (mds *MarketDataService) GetCurrentPrice() core.MarketDataResponse {
	return mds.priceProvider.GetCurrentPrice()
}

// GetPriceHistory returns price history
func (mds *MarketDataService) GetPriceHistory(hours int) []core.MarketDataResponse {
	return mds.priceProvider.GetPriceHistory(hours)
}

// GetPriceChartData returns chart data for prices
func (mds *MarketDataService) GetPriceChartData(hours int) core.ChartDataResponse {
	return mds.chartGenerator.CreatePriceChartData(hours)
}

// GetPnLChartData returns chart data for P&L
func (mds *MarketDataService) GetPnLChartData(positionsData []map[string]interface{}) core.ChartDataResponse {
	return mds.chartGenerator.CreatePnLChartData(positionsData)
}

// GetExposureChartData returns chart data for exposure analysis
func (mds *MarketDataService) GetExposureChartData(exposureData map[string]float64) map[string]interface{} {
	return mds.chartGenerator.CreateExposureChartData(exposureData)
}